aten/src/ATen/native/cuda/MultiTensorApply.cuh

```diff
@@ -95,6 +110,70 @@ struct FusedOptimizerTensorListMetadata {
   int start_tensor_this_launch;
 };
 
+bool can_use_static_tensor_list_meta(
+    std::vector<std::vector<at::Tensor>>& tensor_lists,
+    int depth) {
+  const int64_t n_tensors = tensor_lists[0].size();
+  if (n_tensors > depth_to_max_tensors[depth - 1]) {
+    return false;
+  }
+  int64_t num_blocks = 0;
+  for (const auto t : c10::irange(n_tensors)) {
+    const auto numel = tensor_lists[0][t].numel();
+    const auto chunks = numel / kChunkSize + (numel % kChunkSize != 0);
+    num_blocks += chunks;
+    if (num_blocks > depth_to_max_blocks[depth - 1]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Helper for transfering multiple std::vector<T> onto device with a single
+// page-locked cudaMemcpyAsync.
```

Comment:
For my learning: where is the page locking happening? And the cudaMemcpyAsync? I see a `memcpy` and a `.to` call below--do one of them implicitly call cudaMemcpyAsync?

Reply:
> Update: oooooh is the page-locked referring to the .pinned_memory call below?

Yes this is correct. We get the pinned (a.k.a page-locked) memory from `CachingHostAllocator` so it's practically free. When the src is in pinned memory, `.to(/*non_blocking=*/true)` results in a `cudaMemcpyAsync`.